test/e2e/aks_byo_node.go

```diff
@@ -0,0 +1,190 @@
+					},
+				},
+			},
+			PreKubeadmCommands: []string{"kubeadm init phase upload-config all"},
+		},
+	}
+	err = mgmtClient.Create(ctx, kubeadmConfig)
+	Expect(err).NotTo(HaveOccurred())
+
+	machinePool := &expv1.MachinePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: infraMachinePool.Namespace,
+			Name:      infraMachinePool.Name,
+		},
+		Spec: expv1.MachinePoolSpec{
+			ClusterName: input.Cluster.Name,
+			Replicas:    ptr.To[int32](2),
+			Template: clusterv1.MachineTemplateSpec{
+				Spec: clusterv1.MachineSpec{
+					Bootstrap: clusterv1.Bootstrap{
+						ConfigRef: &corev1.ObjectReference{
+							APIVersion: bootstrapv1.GroupVersion.String(),
+							Kind:       "KubeadmConfig",
+							Name:       kubeadmConfig.Name,
+						},
+					},
+					ClusterName: input.Cluster.Name,
+					InfrastructureRef: corev1.ObjectReference{
+						APIVersion: infrav1.GroupVersion.String(),
+						Kind:       "AzureMachinePool",
+						Name:       infraMachinePool.Name,
+					},
+					Version: ptr.To(input.KubernetesVersion),
+				},
+			},
+		},
+	}
+	err = mgmtClient.Create(ctx, machinePool)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("creating a Kubernetes client to the workload cluster")
+	workloadClusterProxy := bootstrapClusterProxy.GetWorkloadCluster(ctx, input.Cluster.Spec.ControlPlaneRef.Namespace, input.Cluster.Spec.ControlPlaneRef.Name)
+	Expect(workloadClusterProxy).NotTo(BeNil())
+	clientset := workloadClusterProxy.GetClientSet()
+	Expect(clientset).NotTo(BeNil())
+
+	By("Verifying the bootstrap succeeded")
+	Eventually(func(g Gomega) {
+		pool := &infrav1exp.AzureMachinePool{}
+		err := mgmtClient.Get(ctx, client.ObjectKeyFromObject(infraMachinePool), pool)
+		g.Expect(err).NotTo(HaveOccurred())
+		g.Expect(conditions.IsTrue(pool, infrav1.BootstrapSucceededCondition)).To(BeTrue())
+	}, input.WaitIntervals...).Should(Succeed())
+
+	By("Adding the expected AKS labels to the nodes")
+	// TODO: move this to the MachinePool object once MachinePools support label propagation
+	Eventually(func(g Gomega) {
+		nodeList, err := clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+		g.Expect(err).NotTo(HaveOccurred())
+		g.Expect(int32(len(nodeList.Items))).To(Equal(input.ExpectedWorkerNodes + 2))
+		for i, node := range nodeList.Items {
+			if _, ok := node.Labels["kubernetes.azure.com/cluster"]; !ok {
+				node.Labels["kubernetes.azure.com/cluster"] = infraControlPlane.Spec.NodeResourceGroupName
+				_, err := clientset.CoreV1().Nodes().Update(ctx, &nodeList.Items[i], metav1.UpdateOptions{})
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+		}
+	}, input.WaitIntervals...).Should(Succeed())
+
+	By("Verifying the MachinePool becomes ready")
+	Eventually(func(g Gomega) {
+		pool := &expv1.MachinePool{}
+		err := mgmtClient.Get(ctx, client.ObjectKeyFromObject(machinePool), pool)
+		g.Expect(err).NotTo(HaveOccurred())
+		g.Expect(conditions.IsTrue(pool, clusterv1.ReadyCondition)).To(BeTrue())
+	}, input.WaitIntervals...).Should(Succeed())
+}
```

Comment:
The other tests that create MachinePools have been deleting them at the end of the spec. Should we continue that pattern here?

Reply:
I did originally have the delete here. I can add it back. The reason I originally commented it out and then eventually removed it was that when we delete the machinePool before the end of the test we don't collect logs for it which turned out to be useful for debugging. On the plus side by doing the delete in the test we also test delete behavior.

Perhaps a good compromise is to not do the delete in a defer ("no matter what") but only delete the machine pool if everything else in the test passed? I do think it's helpful having logs of the machines and nodes if the test fails.

Reply:
For this one I can see it being valuable to keep around to collect logs more than the other ones. I think I remember trying to implement some kind of "do this if the test passed/failed" in Ginkgo and not seeing a good way to do that. I'm +1 to keeping this test as-is and not deleting the machine pool until the end with everything else.